typesystem/fields.py

```diff
@@ -96,20 +96,23 @@ def __init__(
         trim_whitespace: bool = True,
         max_length: int = None,
         min_length: int = None,
-        pattern: str = None,
+        pattern: typing.Union[str, typing.Pattern] = None,
         format: str = None,
         **kwargs: typing.Any,
     ) -> None:
         super().__init__(**kwargs)
 
         assert max_length is None or isinstance(max_length, int)
         assert min_length is None or isinstance(min_length, int)
-        assert pattern is None or isinstance(pattern, str)
+        assert pattern is None or isinstance(pattern, (str, typing.Pattern))
         assert format is None or isinstance(format, str)
 
         if allow_blank and not self.has_default():
             self.default = ""
 
+        if isinstance(pattern, str):
+            pattern = re.compile(pattern)
```

Comment:
I'd really prefer it if we kept two seperate `.pattern` and `.pattern_regex` attributes.

It's neater in the error message format, and maps more obviously onto corresponding JSON schema arguments.

Alternatively I'd also be happy with a single `.pattern_regex` attribute, and a `@property` for `pattern` that returns `self.pattern_regex.pattern`.

Reply:
the property won't work with the `self.__dict__` approach used for formatting, and `str.format_map()` is also not usuable since it uses subscription (`__getitem__`) instead of attribute lookup.

will go for two attributes.

Reply:
done in follow-up commits. extracted a helper, and massaged the code so that mypy is completely happy as well.